File: hsr-outputs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xinerama.h>

void
print_text(XineramaScreenInfo *si) {
	if (! si) {
		printf("\n");
		return;
	}

	printf("screen_number: %d\n", si->screen_number);
	printf("\tx_org:\t%d\n", si->x_org);
	printf("\ty_org:\t%d\n", si->y_org);
	printf("\twidth:\t%d\n", si->width);
	printf("\theight:\t%d\n", si->height);
}

void
print_csv(XineramaScreenInfo *si) {
	if (! si)
		return;

	printf("%d,%d,%d,%d,%d\n", si->screen_number, si->x_org, si->y_org, si->width, si->height);
}

void
print_json(XineramaScreenInfo *si) {
	if (! si) {
		printf(",\n");
		return;
	}

	printf("  {\n");
	printf("    \"screen\": %d,\n", si->screen_number);
	printf("    \"x\": %d,\n", si->x_org);
	printf("    \"y\": %d,\n", si->y_org);
	printf("    \"width\": %d,\n", si->width);
	printf("    \"height\": %d\n", si->height);
	printf("  }");
}

enum Format { TEXT, CSV, JSON };

int
main(int argc, char **argv)
{
	Display *display = XOpenDisplay(NULL);

	enum Format format = TEXT;
	void (*print)(XineramaScreenInfo*) = print_text;

	if ((argc > 1) && !strcmp(argv[1], "--csv")) {
		print = print_csv;
		format = CSV;
	}
	if ((argc > 1) && !strcmp(argv[1], "--json")) {
		print = print_json;
		format = JSON;
	}

	if (format == CSV)
		printf("screen_number,x_org,y_org,width,height\n");
	if (format == JSON)
		printf("[\n");

	int nscreens = 0;
	XineramaScreenInfo *xisi = XineramaQueryScreens(display, &nscreens);
	for (int i = 0; i < nscreens; i++) {
		if (i)
			print(NULL);
		print(xisi + i);
	}
	XFree(xisi);

	if (format == JSON)
		printf("\n]\n");

	XCloseDisplay(display);

	return 0;
}

File: hsr-outputs.rb
#!/usr/bin/env ruby
# gem install xrandr

require 'json'
require 'xrandr'

outputs = Xrandr::Control.new.outputs.select(&:connected).map do |o|
  x, y = o.position.split('x')
  width, height = o.resolution.split('x')
  { :name => o.name, :x => x, :y => y, :width => width, :height => height }
end

def print_csv(outputs)
  puts("name,x,y,width,height")
  outputs.each { |o| puts("#{o[:name]},#{o[:x]},#{o[:y]},#{o[:width]},#{o[:height]}") }
end

def print_json(outputs)
  puts outputs.to_json
end

def print_text(outputs)
  outputs.each_with_index do |o, i|
    puts('') if i > 0
    puts("name: #{o[:name]}")
    puts("\tx:\t#{o[:x]}")
    puts("\ty:\t#{o[:y]}")
    puts("\twidth:\t#{o[:width]}")
    puts("\theight:\t#{o[:height]}")
  end
end

if ARGV.first == '--csv'
  print_csv(outputs)
elsif ARGV.first == '--json'
  print_json(outputs)
elsif ARGV.first.nil? or ARGV.first == '--text'
  print_text(outputs)
else
  $stderr.puts("syntax: #$0 [<--text|--json|--csv>]")
end

